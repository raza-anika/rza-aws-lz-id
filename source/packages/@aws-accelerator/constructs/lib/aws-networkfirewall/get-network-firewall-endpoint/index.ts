/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as AWS from 'aws-sdk';

import { throttlingBackOff } from '@aws-accelerator/utils';

AWS.config.logger = console;

/**
 * get-network-firewall-endpoint - lambda handler
 *
 * @param event
 * @returns
 */

export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  // Set variables passed in event
  const endpointAz: string = event.ResourceProperties['endpointAz'];
  const firewallArn: string = event.ResourceProperties['firewallArn'];
  const region: string = event.ResourceProperties['region'];
  const solutionId = process.env['SOLUTION_ID'];

  const nfwClient = new AWS.NetworkFirewall({ region: region, customUserAgent: solutionId });

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      let endpointId: string | undefined = undefined;
      const response = await throttlingBackOff(() =>
        nfwClient.describeFirewall({ FirewallArn: firewallArn }).promise(),
      );

      // Check for endpoint in specified AZ
      if (response.FirewallStatus?.SyncStates) {
        endpointId = response.FirewallStatus?.SyncStates[endpointAz].Attachment?.EndpointId;
      }

      if (endpointId) {
        return {
          PhysicalResourceId: endpointId,
          Status: 'SUCCESS',
        };
      } else {
        throw new Error(`Unable to locate Network Firewall endpoint in AZ ${endpointAz}`);
      }

    case 'Delete':
      // Do nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
