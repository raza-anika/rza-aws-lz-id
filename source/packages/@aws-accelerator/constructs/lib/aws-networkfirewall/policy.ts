/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { NfwRuleSourceCustomActionConfig } from '@aws-accelerator/config';

interface INetworkFirewallPolicy extends cdk.IResource {
  /**
   * The Amazon Resource Name (ARN) of the policy.
   */
  readonly policyArn: string;

  /**
   * The ID of the policy.
   */
  readonly policyId: string;

  /**
   * The name of the policy.
   */
  readonly policyName: string;
}

interface StatefulRuleGroupReference {
  resourceArn: string;
  priority?: number;
}

interface StatelessRuleGroupReference {
  priority: number;
  resourceArn: string;
}

export interface FirewallPolicyProperty {
  statelessDefaultActions: string[];
  statelessFragmentDefaultActions: string[];
  statefulDefaultActions?: string[];
  statefulEngineOptions?: string;
  statefulRuleGroupReferences?: StatefulRuleGroupReference[];
  statelessCustomActions?: NfwRuleSourceCustomActionConfig[];
  statelessRuleGroupReferences?: StatelessRuleGroupReference[];
}

interface NetworkFirewallPolicyProps {
  /**
   * The traffic filtering behavior of a firewall policy, defined in a collection of stateless and stateful rule groups and other settings.
   */
  readonly firewallPolicy: FirewallPolicyProperty;

  /**
   * The descriptive name of the firewall policy.
   */
  readonly name: string;

  /**
   * A description of the firewall policy.
   */
  readonly description?: string;

  /**
   * An optional list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];
}

export class NetworkFirewallPolicy extends cdk.Resource implements INetworkFirewallPolicy {
  public readonly policyArn: string;
  public readonly policyId: string;
  public readonly policyName: string;
  private firewallPolicy: cdk.aws_networkfirewall.CfnFirewallPolicy.FirewallPolicyProperty;
  private customActions?: cdk.aws_networkfirewall.CfnFirewallPolicy.CustomActionProperty[];
  private statefulOptions?: cdk.aws_networkfirewall.CfnFirewallPolicy.StatefulEngineOptionsProperty;

  constructor(scope: Construct, id: string, props: NetworkFirewallPolicyProps) {
    super(scope, id);

    // Set initial properties
    this.policyName = props.name;

    // Transform properties as necessary
    if (props.firewallPolicy.statelessCustomActions) {
      this.transformCustom(props.firewallPolicy);
    }
    if (props.firewallPolicy.statefulEngineOptions) {
      this.transformEngineOptions(props.firewallPolicy);
    }

    // Set firewall policy property
    this.firewallPolicy = {
      statelessDefaultActions: props.firewallPolicy.statelessDefaultActions,
      statelessFragmentDefaultActions: props.firewallPolicy.statelessFragmentDefaultActions,
      statefulDefaultActions: props.firewallPolicy.statefulDefaultActions,
      statefulEngineOptions: this.statefulOptions,
      statefulRuleGroupReferences: props.firewallPolicy.statefulRuleGroupReferences,
      statelessCustomActions: this.customActions,
      statelessRuleGroupReferences: props.firewallPolicy.statelessRuleGroupReferences,
    };

    // Set name tag
    props.tags?.push({ key: 'Name', value: this.policyName });

    const resource = new cdk.aws_networkfirewall.CfnFirewallPolicy(this, 'Resource', {
      firewallPolicy: this.firewallPolicy,
      firewallPolicyName: this.policyName,
      description: props.description,
      tags: props.tags,
    });

    this.policyArn = resource.ref;
    this.policyId = resource.attrFirewallPolicyId;
  }

  /**
   * Transform custom actions to conform with L1 construct.
   *
   * @param props
   */
  private transformCustom(props: FirewallPolicyProperty) {
    const property = props.statelessCustomActions;
    this.customActions = [];

    for (const action of property ?? []) {
      this.customActions.push({
        actionDefinition: {
          publishMetricAction: {
            dimensions: action.actionDefinition.publishMetricAction.dimensions.map(item => {
              return { value: item };
            }),
          },
        },
        actionName: action.actionName,
      });
    }
  }

  /**
   * Transform engine options to conform with L1 construct.
   *
   * @param props
   */
  private transformEngineOptions(props: FirewallPolicyProperty) {
    const property = props.statefulEngineOptions;

    if (property) {
      this.statefulOptions = { ruleOrder: property };
    }
  }
}
