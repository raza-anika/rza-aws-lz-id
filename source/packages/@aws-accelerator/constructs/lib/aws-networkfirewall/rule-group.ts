/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { NfwRuleGroupRuleConfig } from '@aws-accelerator/config';

interface INetworkFirewallRuleGroup extends cdk.IResource {
  /**
   * The Amazon Resource Name (ARN) of the rule group.
   */
  readonly groupArn: string;

  /**
   * The ID of the rule group.
   */
  readonly groupId: string;

  /**
   * The name of the rule group.
   */
  readonly groupName: string;
}

interface NetworkFirewallRuleGroupProps {
  /**
   * The maximum operating resources that this rule group can use.
   */
  readonly capacity: number;

  /**
   * The name of the rule group.
   */
  readonly name: string;

  /**
   * Indicates whether the rule group is stateless or stateful.
   */
  readonly type: string;

  /**
   * A description of the rule group.
   */
  readonly description?: string;

  /**
   * An object that defines the rule group rules.
   */
  readonly ruleGroup?: NfwRuleGroupRuleConfig;

  /**
   * An optional list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];
}

export class NetworkFirewallRuleGroup extends cdk.Resource implements INetworkFirewallRuleGroup {
  public readonly groupArn: string;
  public readonly groupId: string;
  public readonly groupName: string;
  private ruleGroup?: cdk.aws_networkfirewall.CfnRuleGroup.RuleGroupProperty;
  private statelessRules?: cdk.aws_networkfirewall.CfnRuleGroup.StatelessRuleProperty[];
  private customActions?: cdk.aws_networkfirewall.CfnRuleGroup.CustomActionProperty[];
  private ruleVariables?: cdk.aws_networkfirewall.CfnRuleGroup.RuleVariablesProperty;
  private ruleOptions?: cdk.aws_networkfirewall.CfnRuleGroup.StatefulRuleOptionsProperty;

  constructor(scope: Construct, id: string, props: NetworkFirewallRuleGroupProps) {
    super(scope, id);

    // Set initial properties
    this.groupName = props.name;

    // Transform properties as necessary
    if (props.ruleGroup) {
      let statelessAndCustom = undefined;
      if (props.ruleGroup.rulesSource.statelessRulesAndCustomActions) {
        this.transformStatelessCustom(props.ruleGroup);
        statelessAndCustom = {
          statelessRules: this.statelessRules!,
          customActions: this.customActions,
        };
      }
      if (props.ruleGroup.ruleVariables) {
        this.transformRuleVariables(props.ruleGroup);
      }
      if (props.ruleGroup.statefulRuleOptions) {
        this.transformRuleOptions(props.ruleGroup);
      }

      // Set rule group property
      this.ruleGroup = {
        rulesSource: {
          rulesSourceList: props.ruleGroup.rulesSource.rulesSourceList,
          rulesString: props.ruleGroup.rulesSource.rulesString,
          statefulRules: props.ruleGroup.rulesSource.statefulRules,
          statelessRulesAndCustomActions: statelessAndCustom,
        },
        ruleVariables: this.ruleVariables,
        statefulRuleOptions: this.ruleOptions,
      };
    }

    // Set name tag
    props.tags?.push({ key: 'Name', value: this.groupName });

    const resource = new cdk.aws_networkfirewall.CfnRuleGroup(this, 'Resource', {
      capacity: props.capacity,
      ruleGroupName: this.groupName,
      type: props.type,
      description: props.description,
      ruleGroup: this.ruleGroup,
      tags: props.tags,
    });

    this.groupArn = resource.ref;
    this.groupId = resource.attrRuleGroupId;
  }

  /**
   * Transform stateless and custom rule group policies to conform with L1 construct.
   *
   * @param props
   */
  private transformStatelessCustom(props: NfwRuleGroupRuleConfig) {
    const property = props.rulesSource.statelessRulesAndCustomActions;
    this.statelessRules = [];
    this.customActions = [];

    // Push stateless rules
    for (const rule of property?.statelessRules ?? []) {
      this.statelessRules.push({
        priority: rule.priority,
        ruleDefinition: {
          actions: rule.ruleDefinition.actions,
          matchAttributes: {
            destinationPorts: rule.ruleDefinition.matchAttributes?.destinationPorts ?? [],
            destinations:
              rule.ruleDefinition.matchAttributes?.destinations?.map(item => {
                return { addressDefinition: item };
              }) ?? [],
            protocols: rule.ruleDefinition.matchAttributes?.protocols ?? [],
            sourcePorts: rule.ruleDefinition.matchAttributes?.sourcePorts ?? [],
            sources:
              rule.ruleDefinition.matchAttributes?.sources?.map(item => {
                return { addressDefinition: item };
              }) ?? [],
            tcpFlags: rule.ruleDefinition.matchAttributes?.tcpFlags,
          },
        },
      });
    }

    // Push custom actions
    for (const action of property?.customActions ?? []) {
      this.customActions.push({
        actionDefinition: {
          publishMetricAction: {
            dimensions: action.actionDefinition.publishMetricAction.dimensions.map(item => {
              return { value: item };
            }),
          },
        },
        actionName: action.actionName,
      });
    }
  }

  /**
   * Transform rule variables to conform with L1 construct.
   *
   * @param props
   */
  private transformRuleVariables(props: NfwRuleGroupRuleConfig) {
    const property = props.ruleVariables;

    if (property) {
      // Create ipset object
      const ipSet: { [key: string]: { definition: string[] } } = {};
      const ipSetKey = property.ipSets.name;
      ipSet[ipSetKey] = { definition: property.ipSets.definition };

      // Create portset object
      const portSet: { [key: string]: { definition: string[] } } = {};
      const portSetKey = property.portSets.name;
      portSet[portSetKey] = { definition: property.portSets.definition };

      // Instantiate ruleVariables object
      this.ruleVariables = {
        ipSets: ipSet,
        portSets: portSet,
      };
    }
  }

  /**
   * Transform rule options to conform with L1 construct.
   *
   * @param props
   */
  private transformRuleOptions(props: NfwRuleGroupRuleConfig) {
    const property = props.statefulRuleOptions;

    if (property) {
      this.ruleOptions = { ruleOrder: property };
    }
  }
}
