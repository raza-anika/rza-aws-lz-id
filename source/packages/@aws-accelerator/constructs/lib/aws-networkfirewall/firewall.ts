/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { GetNetworkFirewallEndpoint } from './get-network-firewall-endpoint';

interface INetworkFirewall extends cdk.IResource {
  /**
   * The unique IDs of the firewall endpoints for all of the subnets that you attached to the firewall.
   */
  readonly endpointIds: string[];

  /**
   * The Amazon Resource Name (ARN) of the firewall.
   */
  readonly firewallArn: string;

  /**
   * The ID of the firewall.
   */
  readonly firewallId: string;

  /**
   * The name of the policy.
   */
  readonly firewallName: string;
}

interface NetworkFirewallProps {
  /**
   * The Amazon Resource Name (ARN) of the firewall policy.
   */
  readonly firewallPolicyArn: string;

  /**
   * The descriptive name of the firewall.
   */
  readonly name: string;

  /**
   * The subnets that Network Firewall is using for the firewall.
   */
  readonly subnets: string[];

  /**
   * The unique identifier of the VPC where the firewall is in use.
   */
  readonly vpcId: string;

  /**
   * A flag indicating whether it is possible to delete the firewall.
   */
  readonly deleteProtection?: boolean;

  /**
   * A description of the firewall.
   */
  readonly description?: string;

  /**
   * A setting indicating whether the firewall is protected against a change to the firewall policy association.
   */
  readonly firewallPolicyChangeProtection?: boolean;

  /**
   * A setting indicating whether the firewall is protected against changes to the subnet associations.
   */
  readonly subnetChangeProtection?: boolean;

  /**
   * An optional list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];
}

export class NetworkFirewall extends cdk.Resource implements INetworkFirewall {
  public readonly endpointIds: string[];
  public readonly firewallArn: string;
  public readonly firewallId: string;
  public readonly firewallName: string;
  private subnetMapping: cdk.aws_networkfirewall.CfnFirewall.SubnetMappingProperty[];

  constructor(scope: Construct, id: string, props: NetworkFirewallProps) {
    super(scope, id);

    // Set initial properties
    this.firewallName = props.name;
    this.subnetMapping = props.subnets.map(item => {
      return { subnetId: item };
    });

    // Set name tag
    props.tags?.push({ key: 'Name', value: this.firewallName });

    const resource = new cdk.aws_networkfirewall.CfnFirewall(this, 'Resource', {
      firewallName: this.firewallName,
      firewallPolicyArn: props.firewallPolicyArn,
      subnetMappings: this.subnetMapping,
      vpcId: props.vpcId,
      deleteProtection: props.deleteProtection,
      description: props.description,
      firewallPolicyChangeProtection: props.firewallPolicyChangeProtection,
      subnetChangeProtection: props.subnetChangeProtection,
      tags: props.tags,
    });

    // Set remaining properties
    this.endpointIds = resource.attrEndpointIds;
    this.firewallArn = resource.ref;
    this.firewallId = resource.attrFirewallId;
  }

  public addLogging(config: cdk.aws_networkfirewall.CfnLoggingConfiguration.LoggingConfigurationProperty) {
    new cdk.aws_networkfirewall.CfnLoggingConfiguration(this, 'LoggingConfig', {
      firewallArn: this.firewallArn,
      loggingConfiguration: config,
    });
  }

  public addNetworkFirewallRoute(
    id: string,
    destination: string,
    endpointAz: string,
    logGroupKmsKey: cdk.aws_kms.Key,
    logRetentionInDays: number,
    routeTableId: string,
  ): void {
    // Get endpoint ID from custom resource
    const vpcEndpointId = new GetNetworkFirewallEndpoint(this, `${id}Endpoint`, {
      endpointAz: endpointAz,
      firewallArn: this.firewallArn,
      kmsKey: logGroupKmsKey,
      logRetentionInDays: logRetentionInDays,
      region: cdk.Stack.of(this).region,
    }).endpointId;

    new cdk.aws_ec2.CfnRoute(this, id, {
      routeTableId,
      destinationCidrBlock: destination,
      vpcEndpointId,
    });
  }
}
