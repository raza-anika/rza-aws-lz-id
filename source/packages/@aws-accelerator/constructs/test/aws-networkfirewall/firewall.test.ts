/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { NetworkFirewall } from '../../lib/aws-networkfirewall/firewall';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(NetworkFirewall): ';

//Initialize stack for resource configuration test
const stack = new cdk.Stack();

const firewallPolicyArn = 'arn:aws:network-firewall:us-east-1:222222222222:firewall-policy/TestPolicy';

new NetworkFirewall(stack, 'TestFirewall', {
  firewallPolicyArn: firewallPolicyArn,
  name: 'TestFirewall',
  subnets: ['Test-Subnet-1', 'Test-Subnet-2'],
  vpcId: 'TestVpc',
  tags: [],
});

/**
 * Network Firewall construct test
 */
describe('Network Firewall', () => {
  snapShotTest(testNamePrefix, stack);
});
