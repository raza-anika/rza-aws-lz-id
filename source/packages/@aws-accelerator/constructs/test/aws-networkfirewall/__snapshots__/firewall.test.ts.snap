// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Network Firewall Construct(NetworkFirewall):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "TestFirewallE26FCA5C": Object {
      "Properties": Object {
        "FirewallName": "TestFirewall",
        "FirewallPolicyArn": "arn:aws:network-firewall:us-east-1:222222222222:firewall-policy/TestPolicy",
        "SubnetMappings": Array [
          Object {
            "SubnetId": "Test-Subnet-1",
          },
          Object {
            "SubnetId": "Test-Subnet-2",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestFirewall",
          },
        ],
        "VpcId": "TestVpc",
      },
      "Type": "AWS::NetworkFirewall::Firewall",
    },
  },
}
`;
